### ISSUE TITLE
Ban peers we can't sync from

### ISSUE THREAD

**Issue by ignopeverell** [mimblewimble/grin#406]
We have a possible DoS where a peer can show up with high work and just not send any header back to our requests or send headers we already have, forever. We won't be able to sync with it and if we need to sync with anyone else, we won't be able either because that peer will be the one we're always picking.

**Comment by ignopeverell** [comment]
Note that this likely requires #402 first.

**Comment by yeastplume** [comment]
I'm seeing some variation of this as well right now, or something very similar, completely preventing sync. I've added a bit more logging locally to trace what's going on. I've also inserted a bit of code to exclude banned peers when getting most_work_peer before the next peer refresh has kicked in.

```
Dec 05 11:36:02.055 TRCE peer diff: 132149772, local diff padded: 131388810
Dec 05 11:36:02.056 INFO sync: late on the most worked chain, enabling sync
Dec 05 11:36:02.056 DEBG Running header sync.
Dec 05 11:36:02.056 DEBG Header head is: Tip { height: 14566, last_block_h: 96b4a94f, prev_block_h: d25c9b51, total_difficulty: Difficulty { num: 133376476 } }
Dec 05 11:36:02.057 DEBG Most work peer is: PeerInfo { capabilities: FULL_HIST | UTXO_HIST | PEER_LIST | FULL_NODE, user_agent: "MW/Grin 0.1", version: 1, addr: V4(37.59.48.28:13414), total_difficulty: Difficulty { num: 132149772 } }
Dec 05 11:36:02.057 DEBG Running body sync.
Dec 05 11:36:02.057 DEBG body_sync: body_head - 3430309d, 13930, header_head - 96b4a94f, 14566, sync_head - 96b4a94f, 14566
Dec 05 11:36:02.124 DEBG block_sync: requesting blocks (13930/14566), [f892f7ce, 7dac6c8a, 3a267e4b, 97d1fd71, d0b5cc9c, 83ac9410, 50bb311d, 801a4839, f8e25890, caaf5104]
Dec 05 11:36:02.124 DEBG Requesting block f892f7ce from peer 37.59.48.28:13414.
Dec 05 11:36:02.124 DEBG Requesting block 7dac6c8a from peer 37.59.48.28:13414.
```

The header height peer we've attached to actually has a lower difficulty than our own header height.. however we still have blocks to download and this is our best peer, so on we go:

```
Dec 05 11:36:02.057 DEBG body_sync: body_head - 3430309d, 13930, header_head - 96b4a94f, 14566, sync_head - 96b4a94f, 14566
Dec 05 11:36:02.124 DEBG block_sync: requesting blocks (13930/14566), [f892f7ce, 7dac6c8a, 3a267e4b, 97d1fd71, d0b5cc9c, 83ac9410, 50bb311d, 801a4839, f8e25890, caaf5104]
Dec 05 11:36:02.124 DEBG Requesting block f892f7ce from peer 37.59.48.28:13414.
Dec 05 11:36:02.124 DEBG Requesting block 7dac6c8a from peer 37.59.48.28:13414.
Dec 05 11:36:02.124 DEBG Requesting block 3a267e4b from peer 37.59.48.28:13414.
Dec 05 11:36:02.124 DEBG Requesting block 97d1fd71 from peer 37.59.48.28:13414.
Dec 05 11:36:02.124 DEBG Requesting block d0b5cc9c from peer 37.59.48.28:13414.
Dec 05 11:36:02.124 DEBG Requesting block 83ac9410 from peer 37.59.48.28:13414.
Dec 05 11:36:02.124 DEBG Requesting block 50bb311d from peer 37.59.48.28:13414.
Dec 05 11:36:02.124 DEBG Requesting block 801a4839 from peer 37.59.48.28:13414.
Dec 05 11:36:02.124 DEBG Requesting block f8e25890 from peer 37.59.48.28:13414.
Dec 05 11:36:02.124 DEBG Requesting block caaf5104 from peer 37.59.48.28:13414.
Dec 05 11:36:03.124 DEBG Running body sync.
Dec 05 11:36:03.125 DEBG body_sync: body_head - 3430309d, 13930, header_head - 96b4a94f, 14566, sync_head - 96b4a94f, 14566
Dec 05 11:36:03.188 DEBG block_sync: requesting blocks (13930/14566), [f892f7ce, 7dac6c8a, 3a267e4b, 97d1fd71, d0b5cc9c, 83ac9410, 50bb311d, 801a4839, f8e25890, caaf5104]
Dec 05 11:36:03.188 DEBG Requesting block f892f7ce from peer 37.59.48.28:13414.
Dec 05 11:36:03.188 DEBG Requesting block 7dac6c8a from peer 37.59.48.28:13414.
Dec 05 11:36:03.188 DEBG Requesting block 3a267e4b from peer 37.59.48.28:13414.
```

This goes on infinitely... because the peer doesn't appear to have any blocks in common. Tweaking the code a bit to do a header sync first further demonstrates this.

**Comment by ignopeverell** [comment]
You may be seeing what was fixed in 9945371, the symptoms were pretty much the same. This issue was that the peer(s) we were trying to get info from had pars of the fork we were on locallly. So we would send our locator and they would find a common block, but that'd be on the losing fork we were trying to get out of. And they actually had even less information about it.

**Comment by ignopeverell** [comment]
Hasn't seemed to be as much of a pb, seems to be more a consequence than a cause. Will file something more specific with testnet2 if we run into similar issues again.
